packages/ssz/src/viewDU/container.ts

```diff
@@ -0,0 +1,222 @@
+import {getNodeAtDepth, LeafNode, Node, setNodesAtDepth} from "@chainsafe/persistent-merkle-tree";
+import {BasicType, Type} from "../type/abstract";
+import {CompositeType, TreeViewDU} from "../type/composite";
+import {ContainerTypeGeneric} from "../view/container";
+
+/* eslint-disable @typescript-eslint/member-ordering */
+
+export type FieldsViewDU<Fields extends Record<string, Type<unknown>>> = {
+  [K in keyof Fields]: Fields[K] extends CompositeType<unknown, unknown, infer TVDU>
+    ? // If composite, return view. MAY propagate changes updwards
+      TVDU
+    : // If basic, return struct value. Will NOT propagate changes upwards
+    Fields[K] extends BasicType<infer V>
+    ? V
+    : never;
+};
+
+export type ContainerTreeViewDUType<Fields extends Record<string, Type<unknown>>> = FieldsViewDU<Fields> &
+  TreeViewDU<ContainerTypeGeneric<Fields>>;
+export type ContainerTreeViewDUTypeConstructor<Fields extends Record<string, Type<unknown>>> = {
+  new (type: ContainerTypeGeneric<Fields>, node: Node, cache?: unknown): ContainerTreeViewDUType<Fields>;
+};
+
+type ContainerTreeViewDUCache = {
+  nodes: Node[];
+  caches: unknown[];
+  nodesPopulated: boolean;
+};
+
+class ContainerTreeViewDU<Fields extends Record<string, Type<unknown>>> extends TreeViewDU<
+  ContainerTypeGeneric<Fields>
+> {
+  protected nodes: Node[] = [];
+  protected caches: unknown[];
+  protected readonly nodesChanged = new Set<number>();
+  protected readonly viewsChanged = new Map<number, unknown>();
+  private nodesPopulated: boolean;
+
+  constructor(
+    readonly type: ContainerTypeGeneric<Fields>,
+    protected _rootNode: Node,
+    cache?: ContainerTreeViewDUCache
+  ) {
+    super();
+
+    if (cache) {
+      this.nodes = cache.nodes;
+      this.caches = cache.caches;
+      this.nodesPopulated = cache.nodesPopulated;
+    } else {
+      this.nodes = [];
+      this.caches = [];
+      this.nodesPopulated = false;
+    }
+  }
+
+  get node(): Node {
+    return this._rootNode;
+  }
+
+  get cache(): ContainerTreeViewDUCache {
+    return {
+      nodes: this.nodes,
+      caches: this.caches,
+      nodesPopulated: this.nodesPopulated,
+    };
+  }
+
+  commit(): Node {
+    if (this.nodesChanged.size === 0 && this.viewsChanged.size === 0) {
+      return this._rootNode;
+    }
+
+    const nodesChanged: {index: number; node: Node}[] = [];
+
+    for (const [index, view] of this.viewsChanged) {
+      nodesChanged.push({
+        index,
+        node: (
+          this.type.fieldsEntries[index].fieldType as unknown as CompositeType<unknown, unknown, unknown>
+        ).commitViewDU(view),
+      });
+    }
+
+    for (const index of this.nodesChanged) {
+      nodesChanged.push({index, node: this.nodes[index]});
+    }
+
+    // TODO: Optimize to loop only once
```

Comment:
implement this TODO ?

Reply:
I don't if Javascript allows a better approach. All ideas I've found online are variations of this